src/server/src/main/java/io/cassandrareaper/service/SegmentGenerator.java

```diff
@@ -81,76 +87,186 @@ static boolean greaterThanOrEqual(BigInteger big0, BigInteger big1) {
+      repairSegments =
+          coalesceTokenRanges(getTargetSegmentSize(totalSegmentCount), replicasToRange);
     }
 
-    // verify that the whole range is repaired
-    BigInteger total = BigInteger.ZERO;
-    for (RingRange segment : repairSegments) {
-      BigInteger size = segment.span(rangeSize);
-      total = total.add(size);
+    return repairSegments;
+  }
+
+  @VisibleForTesting
+  List<Segment> coalesceTokenRanges(
+      BigInteger targetSegmentSize, Map<List<String>, List<RingRange>> replicasToRange) {
+
+    List<Segment> coalescedRepairSegments = Lists.newArrayList();
+    List<RingRange> tokenRangesForCurrentSegment = Lists.newArrayList();
+    BigInteger tokenCount = BigInteger.ZERO;
+
+    for (Entry<List<String>, List<RingRange>> tokenRangesByReplica : replicasToRange.entrySet()) {
+      LOG.info("Coalescing segments for nodes {}", tokenRangesByReplica.getKey());
+      if (!tokenRangesForCurrentSegment.isEmpty()) {
+        coalescedRepairSegments.add(
+            Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
+        tokenRangesForCurrentSegment = Lists.newArrayList();
+      }
+
+      for (RingRange tokenRange : tokenRangesByReplica.getValue()) {
+        if (tokenRange.span(rangeSize).add(tokenCount).compareTo(targetSegmentSize) > 0
+            && !tokenRangesForCurrentSegment.isEmpty()) {
+          // enough tokens in that segment
+          LOG.info(
+              "Got enough tokens for one segment ({}) : {}",
+              tokenCount,
+              tokenRangesForCurrentSegment);
+          coalescedRepairSegments.add(
+              Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
+          tokenRangesForCurrentSegment = Lists.newArrayList();
+          tokenCount = BigInteger.ZERO;
+        }
+
+        tokenCount = tokenCount.add(tokenRange.span(rangeSize));
+        tokenRangesForCurrentSegment.add(tokenRange);
+
+      }
     }
-    if (!total.equals(rangeSize) && !incrementalRepair) {
-      throw new ReaperException("Not entire ring would get repaired");
+
+    if (!tokenRangesForCurrentSegment.isEmpty()) {
+      coalescedRepairSegments.add(
+          Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
     }
-    return repairSegments;
+
+    int coalescedRanges =
+        coalescedRepairSegments
+            .stream()
+            .map(segment -> segment.getTokenRanges().size())
+            .reduce((first, second) -> first + second)
+            .orElse(0);
+
+    int totalRanges =
+        replicasToRange
+            .values()
+            .stream()
+            .map(List::size)
+            .reduce((first, second) -> first + second)
+            .orElse(0);
+
+    LOG.debug("Coalesced ranges/Total ranges : {}/{}", coalescedRanges);
+
+    LOG.debug("Total number of ranges : {}", totalRanges);
+
+    // Check that we haven't left any token range outside of the resulting segments
+    Preconditions.checkArgument(
```

Comment:
or `.checkState(..)` ?? (idk. who's at fault here?), or `assert …` ?

also lines 232-239 should be moved into a separate method, as they're only for the precondition call.

for example this would make more sense when/if as an assert…
``` 
assert hasCoalescedAllTokens(coalescedRepairSegments, replicasToRange) : "Number of coalesced tokens doesn't match with the total number of tokens";
```

Reply:
addressed here: https://github.com/thelastpickle/cassandra-reaper/pull/373/files#diff-282828adcfe6dcc488e490223b8f0a4bR231